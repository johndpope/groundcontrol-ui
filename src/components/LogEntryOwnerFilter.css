.LogEntryOwnerFilter {
  float: right;
  margin-top: -1em;
  max-width: 300px;
}
