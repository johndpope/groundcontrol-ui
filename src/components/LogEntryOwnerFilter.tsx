// Copyright 2019 Stratumn
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import graphql from "babel-plugin-relay/macro";
import React from "react";
import { createFragmentContainer } from "react-relay";
import { Dropdown, DropdownProps } from "semantic-ui-react";

import { LogEntryOwnerFilter_items } from "./__generated__/LogEntryOwnerFilter_items.graphql";

import "./LogEntryOwnerFilter.css";

export interface IProps {
  items: LogEntryOwnerFilter_items;
  ownerId?: string;
  onChange: (values: IProps) => any;
}

export function LogEntryOwnerFilter(props: IProps) {
  const { items, ownerId, onChange } = props;
  const options = items.map(({ id, slug, workspace }) => ({
    key: id,
    text: `${workspace.slug}/${slug}`,
    value: id,
  }));

  const handleChange = (_: React.SyntheticEvent<HTMLElement>, { value }: DropdownProps) => {
    onChange({
      ...props,
      ownerId: value ? value as string : undefined,
    });
  };

  return (
    <Dropdown
      className="LogEntryOwnerFilter"
      placeholder="Choose Project..."
      search={true}
      selection={true}
      clearable={true}
      options={options}
      value={ownerId}
      onChange={handleChange}
    />
  );
}

export default createFragmentContainer(LogEntryOwnerFilter, graphql`
  fragment LogEntryOwnerFilter_items on Project @relay(plural: true) {
    id
    slug
    workspace {
      slug
    }
  }`,
);
