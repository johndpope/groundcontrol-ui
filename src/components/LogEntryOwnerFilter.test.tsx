// Copyright 2019 Stratumn
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { shallow } from "enzyme";
import React from "react";
import { mocked } from "ts-jest/utils";

import { mockQueryPropAttrs } from "../testing/relay";

import { LogEntryOwnerFilter } from "./LogEntryOwnerFilter";

const props = {
  items: [
    {
      ...mockQueryPropAttrs(),
      name: "workspaceName1",
      services: {
        edges: [
          {
            node: {
              id: "serviceId1",
              name: "serviceName1",
            },
          },
        ],
      },
      tasks: {
        edges: [
          {
            node: {
              id: "workspaceId1",
              name: "workspaceName1",
            },
          },
        ],
      },
    },
    {
      ...mockQueryPropAttrs(),
      name: "workspaceName2",
      services: {
        edges: [
          {
            node: {
              id: "serviceId3",
              name: "serviceName3",
            },
          },
        ],
      },
      tasks: {
        edges: [
          {
            node: {
              id: "workspaceId3",
              name: "workspaceName3",
            },
          },
        ],
      },
    },
  ],
  onChange: jest.fn(),
  ownerId: "id2",
  systemId: "systemID",
};

beforeEach(() => {
  mocked(props.onChange).mockClear();
});

describe("<LogEntryOwnerFilter />", () => {
  it("renders items correctly", () => {
    const wrapper = shallow(<LogEntryOwnerFilter {...props} />);
    expect(wrapper).toMatchSnapshot();
  });

  it("triggers onChange when the dropdown value is changed", () => {
    const wrapper = shallow(<LogEntryOwnerFilter {...props} />);
    wrapper.find("Dropdown").simulate("change", null, { value: "id1" });
    expect(props.onChange).toBeCalledTimes(1);
    expect(props.onChange).toBeCalledWith({ ...props, ownerId: "id1" });
  });
});
