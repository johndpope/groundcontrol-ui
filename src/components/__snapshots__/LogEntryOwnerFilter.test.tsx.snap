// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LogEntryOwnerFilter /> renders items correctly 1`] = `
<Dropdown
  additionLabel="Add "
  additionPosition="top"
  className="LogEntryOwnerFilter"
  clearable={true}
  closeOnBlur={true}
  deburr={false}
  fluid={true}
  icon="dropdown"
  minCharacters={1}
  noResultsMessage="No results found."
  onChange={[Function]}
  openOnFocus={true}
  options={
    Array [
      Object {
        "key": "systemID",
        "text": "System",
        "value": "systemID",
      },
      Object {
        "key": "serviceId1",
        "text": "workspaceName1 » serviceName1",
        "value": "serviceId1",
      },
      Object {
        "key": "workspaceId1",
        "text": "workspaceName1 » workspaceName1",
        "value": "workspaceId1",
      },
      Object {
        "key": "serviceId3",
        "text": "workspaceName2 » serviceName3",
        "value": "serviceId3",
      },
      Object {
        "key": "workspaceId3",
        "text": "workspaceName2 » workspaceName3",
        "value": "workspaceId3",
      },
    ]
  }
  placeholder="Choose Namespace..."
  renderLabel={[Function]}
  search={true}
  searchInput="text"
  selectOnBlur={true}
  selectOnNavigation={true}
  selection={true}
  value="id2"
  wrapSelection={true}
/>
`;
